src/app.rs

```diff
@@ -118,133 +82,78 @@ impl App {
-            tf.transform.translation.y + self.initial_bounds[3] / self.zoom,
-        ];
-    }
-
-    pub fn increase_zoom(&mut self) {
-        self.zoom += self.zoom_factor;
-    }
-
-    pub fn decrease_zoom(&mut self) {
-        self.zoom -= self.zoom_factor;
-    }
-
-    pub fn move_pose_estimate(&mut self, x: f64, y: f64, yaw: f64) {
-        let new_yaw = self.pose_estimate.rotation.angle() + yaw;
-        let new_x = x * new_yaw.cos() - y * new_yaw.sin() + self.pose_estimate.translation.x;
-        let new_y = x * new_yaw.sin() + y * new_yaw.cos() + self.pose_estimate.translation.y;
-        self.pose_estimate = Isometry2::new(Vector2::new(new_x, new_y), new_yaw);
+    pub fn run(&mut self) {
+        self.app_modes[self.mode - 1].run();
     }
 
-    pub fn get_pose_estimate(&self) -> rosrust_msg::geometry_msgs::Transform {
-        transformation::iso2d_to_ros(&self.pose_estimate)
-    }
-
-    pub fn deactivate_image_subs(&mut self) {
-        for sub in self.listeners.images.iter_mut() {
-            sub.deactivate();
+    pub fn draw(&self, f: &mut Frame<B>) {
+        if self.show_help {
+            self.show_help(f);
+        } else {
+            self.app_modes[self.mode - 1].draw(f);
         }
     }
 
-    pub fn activate_next_image_sub(&mut self) {
-        let n_subs = &self.listeners.images.len();
-        for (i, sub) in self.listeners.images.iter_mut().enumerate() {
-            if sub.is_active() {
-                sub.deactivate();
-                if i < n_subs - 1 {
-                    let _ = &self.listeners.images[i + 1].activate();
-                } else {
-                    let _ = &self.listeners.images[0].activate();
-                }
-                return;
+    pub fn handle_input(&mut self, input: &String) {
+        if input == app_modes::input::SHOW_HELP {
+            if !self.show_help {
+                self.show_help = true;
+                self.app_modes[self.mode - 1].handle_input(input); // allows apps to do specific actions when help is called
+            } else {
+                self.show_help = false;
             }
         }
-        if n_subs > &0 {
-            let _ = &self.listeners.images[0].activate();
+        if self.show_help {
+            return;
         }
-    }
-
-    pub fn draw_image<B>(&mut self, f: &mut Frame<B>)
-    where
-        B: Backend,
-    {
-        let chunks = Layout::default()
-            .constraints([Constraint::Percentage(100)].as_ref())
-            .split(f.size());
-        for image_sub in &self.listeners.images {
-            if image_sub.is_active() {
-                let image = image_sub.img.read().unwrap();
-                let widget = Image::with_img(image.clone())
-                    .color_mode(ColorMode::Rgb);
-                f.render_widget(widget, chunks[0]);
-                break;
+        match input.trim().parse::<usize>() {
+            Ok(new_mode) => {
```

Comment:
I would vote for keeping the the 'old' key maps. Or give a way to toggle modes. I just hit random numbers until i had the right title. But that might just be a learning curve

Reply:
The new way is good for generating dynamically the help page with the mode switching. The only alternative would be to have inputs SWITCH_MODE_1, SWITCH_MODE_2 and so on and map those as well, but it's basically impossible to reorder modes in the slots right now (or we need to add that to the config file but I think it's an overkill...). What does everybody think?